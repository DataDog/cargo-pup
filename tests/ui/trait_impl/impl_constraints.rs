//@compile-flags: --crate-name test
//@compile-flags: --crate-type lib

// Define traits for testing different aspects of the lint
pub trait MyTrait {
    fn do_something(&self);
}

pub trait AnotherTrait {
    fn another_method(&self);
}

// Test name pattern constraint (should fail)
pub struct BadlyNamedStruct { //~ ERROR: Struct 'BadlyNamedStruct' does not match the required pattern '.*MyTraitImpl$'
    value: String,
}

impl MyTrait for BadlyNamedStruct {
    fn do_something(&self) {
        println!("This implementation has incorrect naming");
    }
}

// Test visibility constraint (should fail)
pub struct VisibilityImpl { //~ ERROR: Struct 'VisibilityImpl' is public, but should be private
    value: String,
}

impl AnotherTrait for VisibilityImpl {
    fn another_method(&self) {
        println!("This implementation has incorrect visibility");
    }
}

// Test compliant implementations (should not fail)
struct GoodMyTraitImpl {
    value: String,
}

impl MyTrait for GoodMyTraitImpl {
    fn do_something(&self) {
        println!("This implementation follows all constraints");
    }
}

struct GoodImpl {
    value: String,
}

impl AnotherTrait for GoodImpl {
    fn another_method(&self) {
        println!("This implementation follows all constraints");
    }
}