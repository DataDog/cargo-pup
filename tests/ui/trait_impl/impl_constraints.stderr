error: Struct 'BadlyNamedStruct' does not match the required pattern '.*MyTraitImpl$'.
  --> tests/ui/trait_impl/impl_constraints.rs:14:1
   |
LL | / pub struct BadlyNamedStruct {
LL | |     value: String,
LL | | }
   | |_^
   |
   = help: Consider renaming the struct.
   = note: Applied by cargo-pup rule 'name_pattern_test'.
   = note: `#[deny(trait_impl)]` on by default

error: Struct 'VisibilityImpl' is public, but should be private.
  --> tests/ui/trait_impl/impl_constraints.rs:25:1
   |
LL | / pub struct VisibilityImpl {
LL | |     value: String,
LL | | }
   | |_^
   |
   = help: Change the visibility to private.
   = note: Applied by cargo-pup rule 'visibility_test'.

error: aborting due to 2 previous errors

